pop-api/examples/fungibles/lib.rs

```diff
@@ -1,108 +1,268 @@
+			self.env().emit_event(Transfer { from: Some(caller), to: Some(to), value });
+			Ok(())
 		}
 
+		/// Transfers `value` tokens on behalf of `from` to the account `to`
+		/// with additional `data` in unspecified format.
+		///
+		/// # Parameters
+		/// - `from` - The account from which the token balance will be withdrawn.
+		/// - `to` - The recipient account.
+		/// - `value` - The number of tokens to transfer.
+		/// - `data` - Additional data with unspecified format.
 		#[ink(message)]
-		pub fn approve(
+		fn transfer_from(
 			&mut self,
-			token: TokenId,
-			spender: AccountId,
+			from: AccountId,
+			to: AccountId,
 			value: Balance,
-		) -> Result<()> {
-			api::approve(token, spender, value)
+			_data: Vec<u8>,
+		) -> Result<(), Psp22Error> {
+			let caller = self.env().caller();
+			// No-op if `from` and `to` is the same address or `value` is zero.
+			if from == to || value == 0 {
+				return Ok(());
+			}
+			// If `from` and the caller are different addresses, a successful transfer results
+			// in decreased allowance by `from` to the caller and an `Approval` event with
+			// the new allowance amount is emitted.
+			api::transfer_from(self.id, from, to, value).map_err(Psp22Error::from)?;
+			self.env().emit_event(Transfer { from: Some(caller), to: Some(to), value });
+			self.env().emit_event(Approval {
+				owner: from,
+				spender: caller,
+				value: self.allowance(from, caller),
+			});
+			Ok(())
+		}
+
+		/// Approves `spender` to spend `value` amount of tokens on behalf of the caller.
+		///
+		/// Successive calls of this method overwrite previous values.
+		///
+		/// # Parameters
+		/// - `spender` - The account that is allowed to spend the tokens.
+		/// - `value` - The number of tokens to approve.
+		#[ink(message)]
+		fn approve(&mut self, spender: AccountId, value: Balance) -> Result<(), Psp22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `spender` is the same address.
+			if caller == spender {
+				return Ok(());
+			}
+			api::approve(self.id, spender, value).map_err(Psp22Error::from)?;
+			self.env().emit_event(Approval { owner: caller, spender, value });
+			Ok(())
 		}
 
+		/// Increases the allowance of `spender` by `value` amount of tokens.
+		///
+		/// # Parameters
+		/// - `spender` - The account that is allowed to spend the tokens.
+		/// - `value` - The number of tokens to increase the allowance by.
 		#[ink(message)]
-		pub fn increase_allowance(
+		fn increase_allowance(
 			&mut self,
-			token: TokenId,
 			spender: AccountId,
 			value: Balance,
-		) -> Result<()> {
-			api::increase_allowance(token, spender, value)
+		) -> Result<(), Psp22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `spender` is the same address or `value` is zero.
+			if caller == spender || value == 0 {
+				return Ok(());
+			}
+			api::increase_allowance(self.id, spender, value).map_err(Psp22Error::from)?;
```

Comment:
This doesnt increase the allowance of the spender for the caller, but increases the allowance of the spender for the contract as that is the origin within the `fungibles::increase_allowance` dispatchable.